File: Frontend/angular-app/src/app/modules/admin/repositories/year/command/update-year-command.ts
import {HttpRequest} from "../../../../../core/services/http/http-request";
import {ServiceResult} from "../../../../../core/models/service-result";
import {Year} from "../../../entities/year";
import {IRequest, IRequestHandler} from "../../../../../core/services/mediator/interfaces";
import {HttpService} from "../../../../../core/services/http/http.service";
import {ApplicationError} from "../../../../../core/exceptions/application-error";
import {MediatorHandler} from "../../../../../core/services/mediator/decorator";
import {NotificationService} from "../../../../../shared/services/notification/notification.service";
import {ValidationRule} from "../../../../../core/validation/validation-rule";
import { Inject } from "@angular/core";

export class UpdateYearCommand extends IRequest<UpdateYearCommand, Year> {
  public id:number | undefined = undefined;
  public companyId:number | undefined = undefined;
  public yearName:number | undefined = undefined;
  public firstDate:Date | undefined = undefined;
  public lastDate:Date | undefined = undefined;
  public isEditable:boolean | undefined = undefined;
  public isCalculable:boolean | undefined = undefined;
  public isCurrentYear:boolean | undefined = undefined;
  public lastEditableDate:Date | undefined = undefined;
  public accountRelationCode:string | undefined = undefined;
  constructor() {
    super();
  }

  mapFrom(entity: Year): UpdateYearCommand {
    this.mapBasics(entity,this);
    return this;
  }

  mapTo(): Year {
    throw new ApplicationError(UpdateYearCommand.name, this.mapTo.name, 'Not Implemented Yet')
  }

  get url(): string {
    return "/admin/year/update";
  }

  get validationRules(): ValidationRule[] {
    return [];
  }
}

@MediatorHandler(UpdateYearCommandHandler.name)
export class UpdateYearCommandHandler implements IRequestHandler<UpdateYearCommand, Year> {
  constructor(
    @Inject(HttpService) private _httpService: HttpService,
    @Inject(NotificationService) private _notificationService: NotificationService,
  ) {
  }

  async Handle(request: UpdateYearCommand): Promise<Year> {
    let httpRequest: HttpRequest<UpdateYearCommand> = new HttpRequest<UpdateYearCommand>(request.url, request);




    return await this._httpService.Put<UpdateYearCommand, ServiceResult<Year>>(httpRequest).toPromise().then(response => {
      this._notificationService.showSuccessMessage()
      return response.objResult
    })

  }
}

File: Backend/Microservices/Admin/Eefa.Admin.Application.Refactor/UseCases/Year/Command/Update/UpdateYearCommand.cs
using System;
using AutoMapper;
using MediatR;
using System.Threading;
using System.Threading.Tasks;

public class UpdateYearCommand : IRequest<ServiceResult<YearModel>>, IMapFrom<Year>
{
    public int Id { get; set; }
    public int YearName { get; set; }
    public DateTime FirstDate { get; set; }
    public DateTime LastDate { get; set; }
    public bool? IsEditable { get; set; }
    public bool IsCalculable { get; set; }
    public DateTime? LastEditableDate { get; set; }
    public bool IsCurrentYear { get; set; } = default!;
    public void Mapping(Profile profile)
    {
        profile.CreateMap<UpdateYearCommand, Year>()
            .IgnoreAllNonExisting();
    }
}


public class UpdateYearCommandHandler : IRequestHandler<UpdateYearCommand, ServiceResult<YearModel>>
{
    private readonly IUnitOfWork _unitOfWork;
    private readonly IMapper _mapper;

    public UpdateYearCommandHandler(IUnitOfWork unitOfWork, IMapper mapper)
    {
        _mapper = mapper;
        _unitOfWork = unitOfWork;
    }

    public async Task<ServiceResult<YearModel>> Handle(UpdateYearCommand request, CancellationToken cancellationToken)
    {
        Year entity = await _unitOfWork.Years.GetByIdAsync(request.Id);

        if (request.IsCurrentYear)
        {
            if (entity.IsCurrentYear == false)
            {
                var entities = await _unitOfWork.Years.GetListAsync();
                foreach (var year in entities)
                {
                    year.IsCurrentYear = false;
                    _unitOfWork.Years.Update(year);
                }
            }
        }

        entity.FirstDate = request.FirstDate;
        entity.LastEditableDate = request.LastEditableDate;
        entity.LastDate = request.LastDate;
        entity.IsEditable = request.IsEditable;
        entity.IsCalculable = request.IsCalculable;
        entity.IsCurrentYear = request.IsCurrentYear;
        entity.YearName = request.YearName;

        _unitOfWork.Years.Update(entity);
        await _unitOfWork.SaveChangesAsync(cancellationToken);

        return ServiceResult.Success(_mapper.Map<YearModel>(entity));
    }
}